File: go/harness/session.go
package harness

import (
	"context"
	"errors"
	"fmt"
	"log/slog"
	"time"

	"github.com/althk/tradekit/go/core/domain"
	"github.com/althk/tradekit/go/core/ports"
	"github.com/althk/tradekit/go/store"
)

// Every process that logs into an Indian broker also wrote the same forty
// lines around BrowserLogin: read yesterday's token from the store, try it,
// fall back to the browser, save what came back. This file is those lines,
// once, over the ports rather than over one adapter.

// SessionBroker is what resuming a session needs of an adapter: the browser
// flow for a new token, the freshness check and installer for a stored one,
// and one cheap authenticated call to find out whether the venue still
// honours it. zerodha, upstox and fyers all satisfy it.
type SessionBroker interface {
	ports.BrowserLogin
	ports.TokenState
	Account(ctx context.Context) (domain.Account, error)
}

// Session says where a token is kept and how a new one is obtained.
type Session struct {
	// Key is the store's kv_state key for the token. Two brokers sharing
	// one store need two keys; the default is "broker_session".
	Key string
	// Callback is where the broker redirects the browser. See BrowserLogin.
	Callback Callback
	// Timeout bounds the browser login. Zero means five minutes: long
	// enough to find the phone for the OTP, short enough that a scheduled
	// run nobody is watching does not hang until the next one.
	Timeout time.Duration
}

// storedSession is the persisted token. Tokens expire overnight, so the issue
// time is what decides whether a stored one is still worth trying.
type storedSession struct {
	AccessToken string    `json:"access_token"`
	IssuedAt    time.Time `json:"issued_at"`
}

// EnsureSession makes broker usable for today: the stored token if it is
// still fresh and the venue accepts it, otherwise a browser login, persisted
// for the next run.
//
// The stored token is checked against the venue and not just by date,
// because a login elsewhere — the broker's web app, another bot — invalidates
// it without changing its age. Only ports.ErrTokenExpired from that check
// leads to the browser; any other failure is returned, since logging in again
// cannot fix a network that is down and would burn the day's login on it.
//
// A token that could not be persisted is not an error: the client already
// holds it, and the worst case is the browser again next run.
func EnsureSession(ctx context.Context, broker SessionBroker, db *store.DB, s Session) error {
	if s.Key == "" {
		s.Key = "broker_session"
	}
	if s.Timeout == 0 {
		s.Timeout = 5 * time.Minute
	}

	var saved storedSession
	err := db.GetState(ctx, s.Key, &saved)
	switch {
	case err == nil && saved.AccessToken != "":
		broker.SetAccessToken(saved.AccessToken, saved.IssuedAt)
		if broker.TokenFresh(ctx) {
			_, err := broker.Account(ctx)
			if err == nil {
				slog.Info("reusing stored broker session", "issued_at", saved.IssuedAt.Format(time.RFC3339))
				return nil
			}
			if !errors.Is(err, ports.ErrTokenExpired) {
				return fmt.Errorf("session: checking stored token: %w", err)
			}
			slog.Info("stored broker session is no longer valid; logging in again")
		}
	case err != nil && !errors.Is(err, store.ErrStateNotFound):
		return fmt.Errorf("session: reading stored token: %w", err)
	}

	loginCtx, cancel := context.WithTimeout(ctx, s.Timeout)
	defer cancel()
	token, err := BrowserLogin(loginCtx, broker, s.Callback)
	if err != nil {
		return fmt.Errorf("session: %w", err)
	}
	if err := db.SetState(ctx, s.Key, storedSession{AccessToken: token, IssuedAt: time.Now()}); err != nil {
		slog.Warn("could not persist the broker session; the next run will need the browser again", "err", err)
	}
	return nil
}

File: py/src/tradekit/harness/session.py
"""Reuse a stored broker token, or log in and store one.

Mirrors ``go/harness/session.go``. Every process that logs into an Indian
broker also wrote the same forty lines around :func:`browser_login`: read
yesterday's token from the store, try it, fall back to the browser, save what
came back. This module is those lines, once, over the ports rather than over
one adapter.
"""

from __future__ import annotations

import datetime as dt
import logging
from dataclasses import dataclass
from typing import Protocol

from tradekit.core import ports
from tradekit.core.domain import Account
from tradekit.store import Database, StateNotFoundError

from .login import DEFAULT_TIMEOUT, Callback, browser_login

__all__ = ["DEFAULT_SESSION_KEY", "Session", "SessionBroker", "ensure_session"]

log = logging.getLogger("tradekit.harness.session")

DEFAULT_SESSION_KEY = "broker_session"


class SessionBroker(ports.BrowserLogin, ports.TokenState, Protocol):
    """What resuming a session needs of an adapter.

    The browser flow for a new token, the freshness check and installer for a
    stored one, and one cheap authenticated call to find out whether the venue
    still honours it. ``UpstoxClient`` and ``FyersClient`` both satisfy it.
    """

    def account(self) -> Account:
        """The funds view; the one cheap call that proves the venue accepts the token."""
        ...


@dataclass(slots=True)
class Session:
    """Where a token is kept and how a new one is obtained.

    Args:
        callback: Where the broker redirects the browser. See :func:`browser_login`.
        key: The store's ``kv_state`` key for the token. Two brokers sharing
            one store need two keys.
        timeout: Bounds the browser login, in seconds. The default is long
            enough to find the phone for the OTP and short enough that a
            scheduled run nobody is watching does not hang until the next one.

    """

    callback: Callback
    key: str = DEFAULT_SESSION_KEY
    timeout: float = DEFAULT_TIMEOUT


def ensure_session(broker: SessionBroker, db: Database, session: Session) -> None:
    """Make ``broker`` usable for today.

    The stored token if it is still fresh and the venue accepts it, otherwise
    a browser login, persisted for the next run.

    The stored token is checked against the venue and not just by date,
    because a login elsewhere -- the broker's web app, another bot --
    invalidates it without changing its age. Only :class:`ports.TokenExpiredError`
    from that check leads to the browser; any other failure propagates, since
    logging in again cannot fix a network that is down and would burn the
    day's login on it.

    A token that could not be persisted is not an error: the client already
    holds it, and the worst case is the browser again next run.
    """
    try:
        saved = db.get_state(session.key)
    except StateNotFoundError:
        saved = None
    if saved and saved.get("access_token"):
        issued_at = dt.datetime.fromisoformat(saved["issued_at"])
        broker.set_access_token(saved["access_token"], issued_at)
        if broker.token_fresh():
            try:
                broker.account()
            except ports.TokenExpiredError:
                log.info("stored broker session is no longer valid; logging in again")
            else:
                log.info("reusing stored broker session issued at %s", saved["issued_at"])
                return

    token = browser_login(broker, session.callback, timeout=session.timeout)
    try:
        db.set_state(
            session.key,
            {"access_token": token, "issued_at": dt.datetime.now(dt.UTC).isoformat()},
        )
    except Exception:  # the token is installed; only the next run is affected
        log.warning("could not persist the broker session; the next run will need the browser again", exc_info=True)
